### ISSUE TITLE
mise tasks which error and cause mise to exit leaves other tasks still running

### ISSUE THREAD

**Issue by saranrapjs** [jdx/mise#2418]
**Describe the bug**
When running multiple long-running mise tasks, when one task exits with an error code, the parent mise process will exit — but other spawned tasks which hadn't exited will continue running in the background.

**To Reproduce**

With a `.mise.toml` like this:
```toml
[tasks.a]
run = "sleep 5 && exit 1"

[tasks.b]
run = "while :\ndo\n  echo waiting && sleep 1\ndone"
```
...running `mise run "*"` will exit after task `a` exits, but you'll continue to see "waiting" in the terminal because task `b` is still running.

**Expected behavior**
When existing the `mise run` process because a sub-process has exited with error, other tasks should be cleaned up.

**`mise doctor` output**

```text
version: 2024.8.2 macos-arm64 (c604cbb 2024-08-01)
activated: yes
shims_on_path: yes

build_info: 
  Target: aarch64-apple-darwin
  Features: DEFAULT, NATIVE_TLS, OPENSSL
  Built: Thu, 1 Aug 2024 11:36:43 +0000
  Rust Version: rustc 1.80.0 (051478957 2024-07-21)
  Profile: release

shell: 
  /bin/zsh
  zsh 5.9 (x86_64-apple-darwin23.0)

dirs: 
  data: ~/.local/share/mise
  config: ~/.config/mise
  cache: ~/Library/Caches/mise
  state: ~/.local/state/mise
  shims: ~/.local/share/mise/shims

config_files: 
  ~/.config/mise/config.toml
  ~/.mise.toml

backends: 
  cargo
  core
  go
  npm
  pipx
  spm
  ubi

plugins: 
  bun         (core)
  deno        (core)
  drone       git@github.com:virtualstaticvoid/asdf-drone.git#1553f4d
  erlang      (core)
  github-cli  git@github.com:bartlomiejdanek/asdf-github-cli.git#e0605b7
  go          (core)
  java        (core)
  jsonnet     git@github.com:Banno/asdf-jsonnet.git#dd36d21
  k6          git@github.com:grimoh/asdf-k6.git#238c54d
  node        (core)
  pnpm        git@github.com:jonathanmorley/asdf-pnpm.git#305baff
  python      (core)
  ruby        (core)
  terraform   git@github.com:asdf-community/asdf-hashicorp.git#f1602b6
  vault       git@github.com:asdf-community/asdf-hashicorp.git#f1602b6
  yarn        git@github.com:twuni/asdf-yarn.git#376c540
  zig         (core)

toolset: 
  go@1.22.0             
  node@20.12.1          
  pnpm@9.5.0            
  yarn@1.22.19          
  k6@0.48.0             
  npm:@nyt/cms-proxy-scoop@8.0.0  

env_vars: 
  MISE_SHELL=zsh

settings: 
  activate_aggressive = false
  all_compile = false
  always_keep_download = false
  always_keep_install = false
  asdf_compat = false
  cargo_binstall = true
  color = true
  disable_default_shorthands = false
  disable_tools = []
  experimental = true
  go_default_packages_file = "~/.default-go-packages"
  go_download_mirror = "https://dl.google.com/go"
  go_repo = "https://github.com/golang/go"
  go_set_gopath = false
  go_set_goroot = true
  go_skip_checksum = false
  http_timeout = 30
  jobs = 4
  legacy_version_file = true
  legacy_version_file_disable_tools = []
  libgit2 = true
  node_compile = false
  not_found_auto_install = true
  paranoid = false
  plugin_autoupdate_last_check_duration = "7d"
  python_default_packages_file = "/Users/jeffsisson/.default-python-packages"
  python_pyenv_repo = "https://github.com/pyenv/pyenv.git"
  raw = false
  trusted_config_paths = []
  quiet = false
  verbose = false
  yes = false
  ci = false
  debug = false
  trace = false
  log_level = "info"
  python_venv_auto_create = false

  [status]
  missing_tools = "if_other_versions_installed"
  show_env = false
  show_tools = false

No warnings found
No problems found

```

**Comment by roele** [comment]
Using a slightly modified `.mise.toml` file i am unable to reproduce the behaviour you are describing.

```toml
[tasks.a]
run = '''
while true; do
  random=$(((RANDOM % 10)+1));
  echo "working on something..."
  if [[ $random -eq 5 ]]; then
    echo "an error occurred";
    exit 1;
  fi;
  sleep 1
done;
'''

[tasks.b]
run = '''
while true; do
  echo "waiting on something...";
  sleep 1;
done;
'''
```

After task `a` exists with error code 1, no further output of task `b` is visible which is what i would expect as exiting the parent process will also stop/kill other child processes.

```shell
✖ mise run "*"
[b] $ while true; do
[a] $ while true; do
[b] waiting on something...
[a] working on something...
[b] waiting on something...
[a] working on something...
[b] waiting on something...
[a] working on something...
[a] working on something...
[a] an error occurred
mise [a] exited with code 1
user@host tasks took 3s
✖ ▒
```

**Comment by saranrapjs** [comment]
When I run your example `.mise.toml`, I still see a zombie process for the "waiting on something" task:

```shell
➜  mise-test mise run '*'
[a] $ while true; do
[b] $ while true; do
waiting on something...
working on something...
waiting on something...
working on something...
waiting on something...
working on something...
working on something...
waiting on something...
working on something...
waiting on something...
working on something...
waiting on something...
an error occurred
mise [a] exited with code 1
➜  mise-test waiting on something...
waiting on something...
waiting on something...
waiting on something...
waiting on something...
```

What operating system are you running on? This behaviors strikes me as something that plausibly differs by OS>

**Comment by roele** [comment]
I am on `aarch64-apple-darwin` as well using `fish` shell. Also tried using `zsh` but cannot reproduce this. 

What Terminal (iTerm2?) are you using?
How does your activation in .zshrc/.zsh_profile look like?

**Comment by saranrapjs** [comment]
I'm using `zsh` (specifically via oh-my-zsh), using the [kitty](https://sw.kovidgoyal.net/kitty/) terminal. In my `~/.zshrc`, I have this:
```
eval "$(/Users/jeffsisson/.local/bin/mise activate zsh)"
```

**Comment by roele** [comment]
Still unable to reproduce with `zsh` and `kitty`. Maybe related to oh-my-zsh? Anything visible (parent process or anything) on process level via `ps` or `htop`?

**Comment by jdx** [comment]
@roele I was able to reproduce this I think and was thinking about a resolution. I don't have my laptop handy but I think I ran some tasks which did sleep and echo after a few seconds, then killed one process which did not kill the rest.

Tracing the code this makes sense because the error handlers in the tasks code exits on the first signal immediately without killing the children.

This unfortunately will require a bit of refactoring because we need to store some state about current running processes.
